src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java

```diff
@@ -250,121 +240,65 @@ public void putDataPoint(String metricName,
+            final DataPointsRowKey dataPointsRowKey = new DataPointsRowKey(metricName, rowTime, dataPoint.getDataStoreDataType(), tags);
+            final ByteBuffer serializedKey = DATA_POINTS_ROW_KEY_SERIALIZER.toByteBuffer(dataPointsRowKey);
             List<Map.Entry<String, Integer>> countsByMetricName = null;
 
             // Write out the row key if it is not cached
-            boolean writeRowKey = false;
-            Boolean isCachedKey = m_rowKeyCache.getIfPresent(rowKey);
-            if (isCachedKey == null) {
-                synchronized (LOCK_ROW_KEY) {
-                    isCachedKey = m_rowKeyCache.getIfPresent(rowKey);
-                    if (null == isCachedKey) {
-                        // leave lock
-                        writeRowKey = true;
-                        m_rowKeyCache.put(rowKey, CACHE_BOOLEAN);
-
-                        int count = ++keyInsertCount;
-                        if(insertCountByMetric.containsKey(metricName)) {
-                            insertCountByMetric.put(metricName, insertCountByMetric.get(metricName) + 1);
-                        }
-                        else {
-                            insertCountByMetric.put(metricName, 1);
-                        }
-
-                        if ((writeTime - keyInsertTimer) > 30_000) {
-                            keyInsertTimer = writeTime;
-                            keyInsertCount = 0;
-                            countsByMetricName = new ArrayList<>();
-                            countsByMetricName.addAll(insertCountByMetric.entrySet());
-                            insertCountByMetric.clear();
-                            logger.warn("RowKeys inserted: count={}", count);
-                        }
-                    }
-                }
-            }
-
-            if (writeRowKey) {
-
-                BoundStatement bs = new BoundStatement(m_psInsertRowKey);
-                bs.setBytes(0, ByteBuffer.wrap(metricName.getBytes(UTF_8)));
-
-                ByteBuffer serializedKey = DATA_POINTS_ROW_KEY_SERIALIZER.toByteBuffer(rowKey);
-
-                bs.setBytes(1, serializedKey);
-                bs.setInt(2, rowKeyTtl);
-                m_session.executeAsync(bs);
-
-                for (String split : m_indexTagList) {
-                    String v = tags.get(split);
-                    if (null == v || "".equals(v)) {
-                        continue;
-                    }
-
-                    bs = new BoundStatement(m_psInsertRowKeySplit);
-                    bs.setString(0, metricName);
-                    bs.setString(1, split);
-                    bs.setString(2, v);
-                    bs.setBytes(3, serializedKey);
-                    bs.setInt(4, rowKeyTtl);
-
-                    m_session.executeAsync(bs);
+            boolean writeRowKeyReverseLookups = false;
+            final boolean rowKeyKnown = rowKeyCache.isKnown(serializedKey);
+            if (!rowKeyKnown) {
```

Comment:
Nice one with getting rid of ` synchronized (LOCK_ROW_KEY)`. 

`serializedKey` is defined by data point's timestamp which gives some assurance that keys won't clash (much). However, under heavy load how risky is going lock-less?

Reply:
all right, the answer is that Caffeine does that